File: Java/Binary-Tree-Inorder-Traversal.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */
public class Solution {
    /**
     * @param root: The root of binary tree.
     * @return: Inorder in ArrayList which contains node values.
     */
    public ArrayList<Integer> inorderTraversal(TreeNode root) {
        // write your code here
        Stack<TreeNode> stk = new Stack<TreeNode>();
        ArrayList<Integer> res = new ArrayList<Integer>();
        if (root == null) return res;
        while (!stk.isEmpty() || root != null) {
            while (root != null) {
                stk.push(root);
                root = root.left;
            }
            if (!stk.isEmpty()) {
                root = stk.peek().right;
                res.add(stk.pop().val);
            }
        }
        return res;
    }
}

File: Cpp/Binary-Tree-Inorder-Traversal.cpp
/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 */
class Solution {
    /**
     * @param root: The root of binary tree.
     * @return: Inorder in vector which contains node values.
     */
public:
    vector<int> inorderTraversal(TreeNode *root) {
        // write your code here
        vector<int> vec;
        if (root == NULL) return vec;
        stack<TreeNode*> st;
        TreeNode *cnt = root;
        while (!st.empty() || cnt != NULL) {
            while (cnt != NULL) {
                st.push(cnt);
                cnt = cnt->left;
            }
            cnt = st.top();
            st.pop();
            vec.push_back(cnt->val);
            cnt = cnt->right;
        }
        return vec;
    }
};
